### ISSUE TITLE
Some question about reference points!

### ISSUE THREAD

**Issue by guwen007** [IDEA-opensource/DAB-DETR#46]
Thank you for your excellent work!  After I read the code about DAB-detr Decoder, I have some question about the reference points. 
In the code of Decoder, "reference_points = new_reference_points.detach()".  I am confused about why should use the detach operator. In my understanding, if using detach operator, the gradient won't backpropagated to the reference embedding, the gradient is cut-off. 
Looking forward to your reply! Thank you!.

**Comment by SlongLiu** [comment]
It is inherited from the deformable detr. We find it works better than un-detached versions.
We can only provide an empirical explanation for this for now, which the operator helps each decoder layer concentrate on its predictions better. 
We also expect a better analysis for this.

**Comment by guwen007** [comment]
> It is inherited from the deformable detr. We find it works better than un-detached versions. We can only provide an empirical explanation for this for now, which the operator helps each decoder layer concentrate on its predictions better. We also expect a better analysis for this.

In my understanding, if using aux_loss in DAB-detr, reference embedding can be optimized from the loss of the first decoder layer( which is not detached). But, in the layer after the first layer,  reference embedding can't be optimizer because of the detach operator.
Do you always using aux_loss?

**Comment by SlongLiu** [comment]
We add aux loss after each decoder layer.

**Comment by guwen007** [comment]
> We add aux loss after each decoder layer.

In that case, reference embedding can be optimized from the loss of the first decoder layer( which is not detached). Besides, bbox_embed can be optimized in each stage.
